src/plugins/kibana_usage_collection/server/collectors/cloud/detector/gcp.ts

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { isString } from 'lodash';
+import { promisify } from 'util';
+import { CloudService, CloudServiceOptions, Request, Response } from './cloud_service';
+import { CloudServiceResponse } from './cloud_response';
+
+// GCP documentation shows both 'metadata.google.internal' (mostly) and '169.254.169.254' (sometimes)
+// To bypass potential DNS changes, the IP was used because it's shared with other cloud services
+const SERVICE_ENDPOINT = 'http://169.254.169.254/computeMetadata/v1/instance';
+
+/**
+ * Checks and loads the service metadata for an Google Cloud Platform VM if it is available.
+ */
+class GCPCloudService extends CloudService {
+  constructor(options: CloudServiceOptions = {}) {
+    super('gcp', options);
+  }
+
+  _checkIfService(request?: Request) {
+    if (!request) {
+      return Promise.reject(new Error('not implemented'));
+    }
+
+    // we need to call GCP individually for each field we want metadata for
+    const fields = ['id', 'machine-type', 'zone'];
+
+    const create = this._createRequestForField;
+    const allRequests = fields.map((field) => promisify(request)(create(field)));
+    return (
+      Promise.all(allRequests)
```

Comment:
Q: Should we aim for the best-effort `Promise.allSettled` to get, at least, something? Or are we aiming for an all-or-nothing approach instead?

Reply:
Good question, this was not an intentional decision as it was just how the service was implemented previously, and TBH I hadn't noticed until you pointed this out 😄 

It would probably be better to collect whatever we can, even if some of the fields fail... though I'd probably vote to do this in a follow-up step.

Reply:
Deal! Let's create a follow up improvement :)